File: Talleres/Taller 3/Problema A - Máquina Misteriosa/maqmist.py
from queue import Queue
from collections import deque

INF = -1

def bfs(vecinos, r):
    n = len(vecinos)
    visitado, d = [False] * n, [INF] * n
    visitado[r], d[r] = True, 0
    Q = Queue()
    Q.put(r)
    while not Q.empty():
        nodo = Q.get()
        for vecino in vecinos[nodo]:
            if not visitado[vecino]:
                visitado[vecino] = True
                d[vecino] = d[nodo] + 1
                Q.put(vecino)
    return d

def main():
    n, m = map(int, input().split())
    if m <= n: 
        min_clicks = n - m
    else:
        V = 2*m
        N_out = [deque() for _ in range(V)]
        for u in range(V):
            if u < m and u > 0:
                N_out[u].extend((u - 1, 2 * u))
            elif u > m:
                N_out[u].append(u - 1)
        d = bfs(N_out, n)
        min_clicks = d[m]
    print(min_clicks)

if __name__ == "__main__":
    main()

File: Talleres/Taller 3/Problema A - Máquina Misteriosa/maqmist.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <list>

using namespace std;

vector<int> bfs(const vector<list<int>>& vecinos, int r) {
	int n = vecinos.size();
	vector<bool> visitado(n, false); vector<int> d(n, -1);
	visitado[r] = true; d[r] = 0;
    queue<int> Q;
    Q.push(r);
    while (!Q.empty()) {
        int u = Q.front(); Q.pop();
        for (int v : vecinos[u]) {
            if (!visitado[v]) {
                visitado[v] = true; d[v] = d[u] + 1;
                Q.push(v);
            }
        }
    }
    return d;
}

int main() {
	int n, m;
	cin >> n >> m;
	int res;
	if (m <= n) res = n - m;
	else {
		int V = 2 * m;
		vector<list<int>> N_out(V);
		for (int u = 0; u < V; u++) {
			if (u < m && u > 0) {
				N_out[u].push_back(u - 1);
				N_out[u].push_back(2 * u);
			} else if (u > m) N_out[u].push_back(u - 1);
		}
		auto dist = bfs(N_out, n);
		res = dist[m];
	}
	cout << res << endl;
	return 0;
}